vectortile/src/main/scala/geotrellis/vectortile/protobuf/ProtobufTile.scala

```diff
@@ -0,0 +1,293 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.vectortile.protobuf
+
+import geotrellis.vector._
+import geotrellis.vectortile.{Layer, VectorTile}
+import geotrellis.vectortile.protobuf.internal._
+import geotrellis.spark.SpatialKey
+import geotrellis.spark.tiling.LayoutDefinition
+import geotrellis.vectortile.protobuf.internal.{vector_tile => vt}
+import geotrellis.vectortile.protobuf.internal.vector_tile.Tile.GeomType.{POINT, LINESTRING, POLYGON}
+
+import scala.collection.mutable.ListBuffer
+
+// --- //
+
+/**
+  * A concrete representation of a VectorTile, as one decoded from Protobuf
+  * bytes. This is the original/default type for VectorTiles.
+  *
+  * {{{
+  * import geotrellis.vectortile.protobuf._
+  *
+  * val bytes: Array[Byte] = ...  // from some `.mvt` file
+  * val key: SpatialKey = ...  // preknown
+  * val layout: LayoutDefinition = ...  // preknown
+  *
+  * val tile: VectorTile = ProtobufTile.fromBytes(bytes, key, layout)
+  * }}}
+  *
+  * @constructor This is not meant to be called directly. See this class's
+  * companion object for the available helper methods.
+  */
+case class ProtobufTile(
+  layers: Map[String, ProtobufLayer]
+) extends VectorTile {
+  /** Encode this VectorTile back into a mid-level Protobuf object. */
+  def toProtobuf: vt.Tile =
+    vt.Tile(layers = layers.values.map(_.toProtobuf).toSeq)
+
+  /** Encode this VectorTile back into its original form of Protobuf bytes. */
+  def toBytes: Array[Byte] = toProtobuf.toByteArray
+}
+
+object ProtobufTile {
+  /** Create a ProtobufTile masked as its parent trait. */
+  def fromPBTile(
+    tile: vt.Tile,
+    key: SpatialKey,
+    layout: LayoutDefinition
+  ): (SpatialKey, VectorTile) = {
+    val tileExtent = layout.mapTransform(key)
+
+    val layers: Map[String, ProtobufLayer] = tile.layers.map({ l =>
+      val pbl = ProtobufLayer(l, tileExtent)
+
+      pbl.name -> pbl
+    }).toMap
+
+    (key, new ProtobufTile(layers))
+  }
+
+  /** Create a [[VectorTile]] from raw Protobuf bytes.
+    *
+    * @param bytes  Raw Protobuf bytes from a `.mvt` file or otherwise.
+    * @param key    A key representing where in the Tile grid ''this'' Tile sits.
+    * @param layout A description of the parent Extent and Tile grid.
+    */
+  def fromBytes(
+    bytes: Array[Byte],
+    key: SpatialKey,
+    layout: LayoutDefinition
+  ): (SpatialKey, VectorTile) = {
+    fromPBTile(vt.Tile.parseFrom(bytes), key, layout)
+  }
+}
+
+/**
+  * A [[Layer]] decoded from Protobuf data. All of its Features are decoded
+  * lazily, making for very fast extraction of single features/geometries.
+  *
+  */
+// TODO The Layer storing `key` and `layout` is likely very wasteful
+// it terms of memory.
+case class ProtobufLayer(
```

Comment:
We should follow the "one file to one class/object/trait combo" rule here and other places.